src/Dynamic/Statistics/Statistics.tsx

```diff
@@ -0,0 +1,462 @@
+        ids.push(id as Snowflake);
+      }
+    }
+    return ids;
+  }, [guildStats]);
+
+  const memberEntries = usePoolEntities({
+    type: "member",
+    guildId: guild.id,
+    ids: allMemberIds,
+  });
+  const membersMap = useMemo(() => {
+    const members: Map<Snowflake, Member> = new Map();
+    for (const memberEntry of memberEntries) {
+      if (memberEntry.isLoaded && memberEntry.entity.isDefined()) {
+        members.set(memberEntry.entity.get.id, memberEntry.entity.get);
+      }
+    }
+    return members;
+  }, [memberEntries]);
+
+  const channelsMap = useMemo(() => {
+    const map: Map<string, Channel> = new Map();
+    for (const channel of channels) {
+      map.set(channel.id as string, channel);
+    }
+    return map;
+  }, [channels]);
+
+  if (currentUser.isDefined())
+    return (
+      <Statistics
+        members={membersMap}
+        channels={channelsMap}
+        currentUser={currentUser.get}
+        isArchitusAdmin={false}
+        dispatch={dispatch}
+        stats={guildStats}
+        {...pageProps}
+      />
+    );
+
+  return null;
+};
+
+export default StatisticsProvider;
+
+const Statistics: React.FC<StatisticsProps> = (props) => {
+  const { guild, stats, currentUser, members, channels } = props;
+
+  const getMemberCount = (): number => {
+    return stats.isDefined() ? stats.get.members.count : 0;
+  };
+
+  const getMessageCount = (): number => {
+    return stats.isDefined() ? stats.get.messages.count : 0;
+  };
+
+  const getChannelData = (): any[] => {
+    const data: any[] = [];
+    if (stats.isDefined()) {
+      const channelIds = stats.get.messages.channels;
+      Object.entries(channelIds).forEach(([key, value]) => {
+        const channel = channels.get(key)
+        if (isDefined(channel)) {
+          data.push({ name: channel.name, count: value });
+        }
+      });
+    }
+    data.sort((a, b) => (a.count < b.count ? 1 : -1));
+    return data;
+  };
+
+  const getMemberData = (): any[] => {
+    const data: any[] = [];
+    if (stats.isDefined()) {
+      const memberIds = stats.get.messages.members;
+      Object.entries(memberIds).forEach(([key, value]) => {
+        const member = members.get(key as Snowflake);
+        if (isDefined(member)) {
+          data.push({ name: member.name, count: value });
+        }
+      });
+    }
+    data.sort((a, b) => (a.count < b.count ? 1 : -1));
+    return data;
+  };
+
+  const getPersonalMessageData = (): any[] => {
+    if (stats.isDefined()) {
+      const total = stats.get.messages.count;
+      const userCount = stats.get.messages.members[currentUser.id as string];
+      return [
+        { name: "me", value: userCount },
+        { name: "not me", value: total - userCount },
+      ];
+    }
+    return [];
+  };
+
+  const getArchitusMessageCount = (): number => {
+    if (stats.isDefined()) {
+      return stats.get.messages.members["448940980218101795"];
+    }
+    return 0;
+  };
+
+  return (
```

Comment:
you should probably split this into a few smaller components so it's easier to manage